web3/_utils/request.py

```diff
@@ -47,10 +47,18 @@ def get_default_http_endpoint() -> URI:
 
 
 def cache_and_return_session(
-    endpoint_uri: URI, session: requests.Session = None
+    endpoint_uri: URI, provider_id: int = None, session: requests.Session = None
 ) -> requests.Session:
     # cache key should have a unique thread identifier
-    cache_key = generate_cache_key(f"{threading.get_ident()}:{endpoint_uri}")
+    if provider_id:
```

Comment:
Should we even provide a choice here? It seems like a unique `provider_id`, and in turn, `cache_key`, should be a requirement otherwise we leave the door open to possible session sharing.